field/subfields.go

```diff
@@ -0,0 +1,223 @@
+package field
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sort"
+
+	"github.com/moov-io/iso8583/encoding"
+	"github.com/moov-io/iso8583/padding"
+)
+
+var _ Field = (*Subfields)(nil)
+
+// Subfields is a wrapper object designed to hold ISO8583 subfields. It packs
+// and unpacks subfields ordered by ID. The following is not supported by
+// Subfields:
+//  - Padding
+//  - Encoding
+//
+// Responsibility for this is delegated recursively to the subfields
+// themselves.
+type Subfields struct {
+	spec *Spec
+
+	orderedFieldIDs []int
+	idToFieldMap    map[int]Field
+
+	data interface{}
+}
+
+// NewSubfields creates a new instance of the *Subfield struct,
+// validates and sets its Spec before returning it.
+// Refer to SetSpec() for more information on Spec validation.
+func NewSubfields(spec *Spec) *Subfields {
+	f := &Subfields{}
+	f.SetSpec(spec)
+	return f
+}
+
+// Spec returns the receiver's spec.
+func (f *Subfields) Spec() *Spec {
+	return f.spec
+}
+
+// SetSpec validates the spec and creates new instances of Fields defined
+// in the specification.
+// NOTE: Subfields do not support encoding and padding. Therefore, users should
+// only pass None or nil values for these types. Passing any other value will
+// result in a panic.
+func (f *Subfields) SetSpec(spec *Spec) {
+	if err := validateSubfieldsSpec(spec); err != nil {
+		panic(err)
+	}
+	f.spec = spec
+	f.idToFieldMap = spec.CreateMessageFields()
+	f.orderedFieldIDs = orderedKeys(f.idToFieldMap)
+}
+
+// SetData traverses through fields provided in the data parameter matches them
+// with their spec definition and calls SetData(...) on each spec field with the
+// appropriate data field.
+//
+// A valid input is as follows:
+//
+//		type SubfieldData struct {
+//			F1 *String
+//			F2 *String
+//			F3 *Numeric
+//		}
+//
+func (f *Subfields) SetData(data interface{}) error {
+	f.data = data
+
+	if f.data == nil {
+		return nil
+	}
+
+	str := reflect.ValueOf(data)
+	strKind := str.Kind()
+	if strKind == reflect.Ptr || strKind == reflect.Interface {
+		// get the struct
+		str = str.Elem()
+	}
```

Comment:
```suggestion
	if str.Kind() == reflect.Ptr || str.Kind() == reflect.Interface {
		// get the struct
		str = str.Elem()
	}
```
otherwise we leave `strKind` hanging around, which is no longer accurate